[MM-27936]: Allow resizing the channel sidebar and right-hand sidebar

kyeongsoosoo opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
- make `lhs` and `rhs` resizable
- add `size` property to sidebar store
#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->

Fixes https://github.com/mattermost/mattermost-server/issues/21675

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |

-->

For window widths <900:

not resizable

For window widths 900-1199:


https://user-images.githubusercontent.com/59413880/209613520-4defe4d2-8cd4-47ac-9b3c-56929e8ed2a1.mov


For window widths 1200-1679:

https://user-images.githubusercontent.com/59413880/209613621-411d0279-c16b-4bdf-b430-4130b41d6d43.mov


For window widths >= 1680px:


https://user-images.githubusercontent.com/59413880/209613761-d9d2c49b-144f-4def-b784-1f4c8776ba0b.mov



#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Allow resizing the channel sidebar and right-hand sidebar
```

kyeongsoosoo committed:
add size property to sidebar store

kyeongsoosoo committed:
set sidebar size when window resized

kyeongsoosoo committed:
make lhs resizable

kyeongsoosoo committed:
make rhs resizable

kyeongsoosoo committed:
add condition

kyeongsoosoo committed:
change variable name

kyeongsoosoo committed:
use requestAnimationFrame

kyeongsoosoo committed:
fix width

kyeongsoosoo committed:
update snapshot

kyeongsoosoo committed:
fix test

mattermod commented:
Hello @kyeongsoosoo,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

mattermod commented:
E2E tests not automatically triggered, because PR has no approval yet. Please ask a developer to review and then try again to attach the QA label.

kyeongsoosoo committed:
fix lint

AshishDhama commented:
@kyeongsoosoo Thanks. A heads up, many people are out for the holidays, so there will be some delay in reviews.

matthewbirtch reviewed (changes requested):
Nice work on this @kyeongsoosoo! I'm really excited to get this into the product finally! A couple notes:

1. The snapping seems to snap to the minimum width, but it should actually snap to the default width. (see [this note in the confluence doc](https://mattermost.atlassian.net/wiki/spaces/GLOAB/pages/2196504587/Resizable+Panels#Snapping-to-the-default))
2. Can we make sure the resize controls appear when within 8px of either side of the division? Right now, it seems like it's only appearing if I'm within 8px of the left side of the division.
3. For the double-click behavior, it looks like it animates back to the default width in the RHS, but not for th LHS. It should be consistent. And I don't think we need the animation at all.
4. When I expand the LHS beyond a certain point, the highlighted channel's background doesn't span the full width
    <img width="452" alt="Screenshot 2023-01-01 at 3 28 24 PM" src="https://user-images.githubusercontent.com/2040554/210183783-1d68d753-23c6-4fc1-a36e-fcabda6330c3.png">
5. In the RHS, if I use the expand button, it expands properly, but it seems like it won't collapse now.
    ![2023-01-01 15 31 56](https://user-images.githubusercontent.com/2040554/210183858-f5255ca2-14b4-4538-8054-840dfb1d0bcf.gif)
6. When the RHS is resized below 375px, we get some collisions of icons in the formatting toolbar in the reply input box. Two options: 1) change the min width to 380px, or 2) make changes to the formatting toolbar to hide more icons under the ••• icon when it gets below 380px.
    <img width="397" alt="Screenshot 2023-01-01 at 3 35 18 PM" src="https://user-images.githubusercontent.com/2040554/210183957-ea4c900e-d81f-46d8-b890-f8545bc3dfca.png">
     
7. If I refresh the page, the width gets reset entirely. The widths should persist on refresh.

kyeongsoosoo committed:
expand width

kyeongsoosoo committed:
remove animation when resizeRHSLine double-clicked

kyeongsoosoo committed:
fix sidebarChannel size

kyeongsoosoo committed:
make rhs collapse properly

kyeongsoosoo committed:
add min mode to formatting toolbar

kyeongsoosoo committed:
save sidebar width to localstorage

kyeongsoosoo committed:
snap the panel when the width approaches the default

kyeongsoosoo commented:
Thanks for the review, @matthewbirtch. I missed some parts.
I fixed all the issues. 

>  Two options: 1) change the min width to 380px, or 2) make changes to the formatting toolbar to hide more icons under the ••• icon when it gets below 380px.

I chose the second option. If the other option is better, I'll fix it again.

<img width="263" alt="스크린샷 2023-01-03 오후 11 31 30" src="https://user-images.githubusercontent.com/59413880/210377502-34fbc992-8be7-413f-b7e3-96897fcfa6ce.png">

matthewbirtch commented:
Thanks for resolving most of those @kyeongsoosoo. It's looking great.

> 1. The snapping seems to snap to the minimum width, but it should actually snap to the default width. (see [this note in the confluence doc](https://mattermost.atlassian.net/wiki/spaces/GLOAB/pages/2196504587/Resizable+Panels#Snapping-to-the-default))

This looks fixed, however, the default width for the RHS changes depending on the screen width. For screens >= 1680px, the default width for the RHS is 500px. For screens < 1680 and > 768, the default width for the RHS is 400px

> 2. Can we make sure the resize controls appear when within 8px of either side of the division? Right now, it seems like it's only appearing if I'm within 8px of the left side of the division.

✅ Fixed

> 3. For the double-click behavior, it looks like it animates back to the default width in the RHS, but not for th LHS. It should be consistent. And I don't think we need the animation at all.

✅ Fixed

> 4. When I expand the LHS beyond a certain point, the highlighted channel's background doesn't span the full width

✅ Fixed

> 5. In the RHS, if I use the expand button, it expands properly, but it seems like it won't collapse now.

✅ Fixed

> 6. When the RHS is resized below 375px, we get some collisions of icons in the formatting toolbar in the reply input box. Two options: 1) change the min width to 380px, or 2) make changes to the formatting toolbar to hide more icons under the ••• icon when it gets below 380px.

✅ Fixed. I'm good with this option.

> 7. If I refresh the page, the width gets reset entirely. The widths should persist on refresh.

✅ Fixed

**New**

8. While viewing at 1920 screen width, I resized the RHS to it's max width. Then I resized the browser width to 1100px and noticed I have an RHS that is too large and not resizable. I guess we need to do some checks when the browser width changes to make sure the RHS width adjusts properly.

9. I noticed that sometimes when I open the RHS it animated the width as it opens. Can we remove this animation? See GIF here:
<img src="https://user-images.githubusercontent.com/2040554/210400580-eff3cbab-ddd3-43a7-a3a2-18627b639ba4.gif" width="600" >

kyeongsoosoo committed:
change xlarge-rhs default width

kyeongsoosoo committed:
handle browser resize

kyeongsoosoo committed:
prevent animation when rhs is opened

kyeongsoosoo commented:
@matthewbirtch I've fixed the remaining issues.
> This looks fixed, however, the default width for the RHS changes depending on the screen width. For screens >= 1680px, the default width for the RHS is 500px. For screens < 1680 and > 768, the default width for the RHS is 400px

Fixed!

> 8. While viewing at 1920 screen width, I resized the RHS to it's max width. Then I resized the browser width to 1100px and noticed I have an RHS that is too large and not resizable. I guess we need to do some checks when the browser width changes to make sure the RHS width adjusts properly.

If the width is not between max and min when the window size changes, width returns to default.

> 9. I noticed that sometimes when I open the RHS it animated the width as it opens. Can we remove this animation? See GIF here

Fixed!

matthewbirtch commented:
Looking great @kyeongsoosoo. I wanted to test this out with the `Threads` view when Collapsed Reply Threads are enabled, but I can't seem to enable that on this test server. @jgilliam17 any ideas why CRT would not be available as a setting in the system console?

jgilliam17 commented:
@matthewbirtch I don't see the flag anymore and the setting for CRT is missing in SysConsole Site Configuration>Posts 
@vish9812 Can you please take a look if the CRT feature flag removal had some unintended consequences? Thanks

jgilliam17 commented:
/update-branch

mattermod commented:
Error trying to update the PR.
Please do it manually.

jgilliam17 commented:
@kyeongsoosoo Can you please rebase branch with master? CRT flag was removed after this PR was created, rebasing may fix the issue with missing CRT - Thanks

kyeongsoosoo committed:
Merge branch 'master' of https://github.com/mattermost/mattermost-webapp into MM-27936_resize_sidebar

kyeongsoosoo commented:
@jgilliam17 branch updated!

esethna commented:
Hmm, test server creation seems to be failing. Thanks for working on this @kyeongsoosoo!

kyeongsoosoo committed:
Merge branch 'master' of https://github.com/mattermost/mattermost-webapp into MM-27936_resize_sidebar

kyeongsoosoo commented:
@esethna I updated the branch again!

matthewbirtch reviewed (approved):
Great work on this @kyeongsoosoo! I'll leave it to QA and dev reviews to find anything else.

esethna commented:
This is awesome, thanks @kyeongsoosoo! 

@matthewbirtch one improvement to consider, should we put a slight delay on the blue highlight effect, so that you can move your mouse over the sidebar or RHS quickly without triggering it. As an example, Slack only triggers the highlight effect if you hover your mouse over the grab area for more than 1 second, otherwise it's just a cursor change so it's less distracting.

matthewbirtch commented:
> @matthewbirtch one improvement to consider, should we put a slight delay on the blue highlight effect, so that you can move your mouse over the sidebar or RHS quickly without triggering it. As an example, Slack only triggers the highlight effect if you hover your mouse over the grab area for more than 1 second, otherwise it's just a cursor change so it's less distracting.

Great suggestion @esethna. Let's do that

kyeongsoosoo committed:
add delay to hover highlight effect

kyeongsoosoo commented:
@esethna  I put a delay on the highlight effect. Can you please check?

https://user-images.githubusercontent.com/59413880/212336658-bb63732b-2155-4f13-bbe0-a61d9a935714.mov

esethna reviewed (approved):
Looks great @kyeongsoosoo!!

Willyfrog commented:
reassigning to @M-ZubairAhmed as I think someone from webplatform should take a look at the changes

hmhealey commented on components/resizable_sidebar/resizable_lhs/index.tsx:
```diff
@@ -0,0 +1,171 @@
+        document.body.style.cursor = 'col-resize';
+
+        if (lhsRef.current) {
+            lhsRef.current.classList.add('sidebar--left-dragged');
```
I have the same question here regarding setting this class using a ref. It seems more idiomatic to pass this below using  the `className` prop instead of setting it via a ref here

kyeongsoosoo replied:
I changed it to use `className` unless using external state is available.

hmhealey commented on components/resizable_sidebar/resizable_rhs/index.tsx:
```diff
@@ -0,0 +1,215 @@
+            rhsRef.current.style.width = `${defaultWidth}px`;
+            rhsRef.current.classList.add('sidebar--right-double-clicked');
+
+            setTimeout(() => {
```
I think we might want to use `requestAnimationFrame` here instead of `setTimeout` as well as another place a few lines below this.

This feels like one of the few times where that name makes sense with how we're using it because we're actually trying to skip an animation here 😀

kyeongsoosoo replied:
I also think it's a perfect case to use :)

hmhealey commented on components/resizable_sidebar/resizable_rhs/index.tsx:
```diff
@@ -0,0 +1,215 @@
+    children: React.ReactNode;
+}
+
+function Resizable({
```
Would it be possible to share more of this code between the LHS and RHS? It would be nice to have a more generic `Resizable` component that we might be able to use in the future elsewhere, but that's certainly not necessary if it ends up being too difficult. I'd rather have two pieces of similar code than one full of a ton of branching logic that's hard to read.

From what I can see, the biggest differences between the two code-wise are where the default/max/min values come from, where the actual width is stored, and the extra overlapping logic for the RHS. Also, there's the fact that you're dragging from the left or right of the sidebar, but I don't know how much the code actually cares about that

kyeongsoosoo replied:
I tried to make a base `Resizable` component.  9e5ea4cfbd7f0a68ef7b445ebfe7e199f49879c0

hmhealey replied:
Awesome, thanks. I'll take a look at the component shortly

hmhealey commented on components/root/root.tsx:
```diff
@@ -222,7 +228,13 @@ export default class Root extends React.PureComponent<Props, State> {
         this.tabletMediaQuery = window.matchMedia(`(min-width: ${Constants.MOBILE_SCREEN_WIDTH + 1}px) and (max-width: ${Constants.TABLET_SCREEN_WIDTH}px)`);
         this.mobileMediaQuery = window.matchMedia(`(max-width: ${Constants.MOBILE_SCREEN_WIDTH}px)`);
 
+        this.smallSidebarMediaQuery = window.matchMedia(`(max-width: ${Constants.SMALL_SIDEBAR_BREAKPOINT}px)`);
```
To avoid cluttering up the Root component with more stuff, would you mind pulling this out into a separate component that's just responsible for updating this state while rendering nothing? We probably should've done the same for the code that watches for when the window is resized, but that's existing code so need to change that now.

You could then add that component to the `render` method below inside the `CompassThemeProvider` that also contains the `SidebarRight` so that it only runs while one of the sidebars is visible. The LHS is part of a child component of that, so we just need to make sure that the RHS exists to track that

kyeongsoosoo replied:
`MediaQuery` related to `Sidebarbar` has been moved. Should we move the existing `MediaQuery` as well?

hmhealey replied:
That's okay where it is. This PR's already rather large, so we don't need to do it here. We're also planning on doing some cleanup which will move logic out of the `Root` component in the next few months, so we can tackle it then

hmhealey commented on components/sidebar/sidebar.tsx:
```diff
@@ -222,7 +224,7 @@ export default class Sidebar extends React.PureComponent<Props, State> {
 
         return (
-            <div
+            <Resizable
```
Would you mind renaming the `Resizable` components to `ResizableLhs` and `ResizableRhs` so that we can tell them apart?

hmhealey commented on components/sidebar_right/sidebar_right.tsx:
```diff
@@ -255,14 +258,20 @@ export default class SidebarRight extends React.PureComponent<Props, State> {
-                <div className={'sidebar--right sidebar--right--width-holder'}/>
                 <div
+                    className={'sidebar--right sidebar--right--width-holder'}
+                    ref={this.sidebarRightWidthHolder}
```
Are you meaning to use the same ref for both this div and the `Resizable` below?

kyeongsoosoo replied:
I tried to change the width of the `sidebarRightWidthHolder` by delivering a ref.

kyeongsoosoo replied:
I wanted to deliver the ref of `div` to `ResizableRhs` using `forwardRef`, so I thought that the element that ref points to is only `div` and both `refs` are necessary.

hmhealey replied:
Ooh okay, I see what the confusion is. I thought the `ref` prop of `ResizableRhs` meant that `sidebarRightWidthHolder` was holding a reference to the `ResizableRhs`, not that you were passing it into the `ResizableRhs` so that it would have access to the `div`.

Instead of having the prop to `ResizableRhs` being called `ref`, could you name it something like `containerRef`?

A prop called just `ref` generally means that the reference is pointing to that component. Similarly, you don't need `forwardRef` for `ResizbleRhs` because `forwardRef` is used to make the reference point to a child of that component rather than to the component itself. For example, we use `forwardRef` in [LocalizedInput](https://github.com/mattermost/mattermost-webapp/blob/master/components/localized_input/localized_input.tsx) so that someone passing a `ref` prop into it gets a reference to the `input` rather than the `Localizedinput` itself

hmhealey commented on sass/base/_structure.scss:
```diff
@@ -168,7 +168,7 @@ body.app__body #root {
             #sidebar-right {
                 position: relative;
-                width: 100%;
+                width: 100% !important;
```
There's quite a few uses of `!important` here which is a bit concerning. Are they actually needed?

kyeongsoosoo replied:
In the process of refactoring, all of these were removed. :)

hmhealey commented on sass/layout/_sidebar-left.scss:
```diff
@@ -23,7 +23,51 @@ $sidebarOpacityAnimationDuration: 0.15s;
+
+    #resizeLine {
+        &::after {
+            background-color: var(--denim-sidebar-active-border);
```
@matthewbirtch Are these supposed to use a theme-specific colour, or should they be based on the theme?

AshishDhama replied:
No, it is not, right color variable is `--sidebar-text-active-border`

hmhealey reviewed (changes requested):
Thanks for the PR! I've got a bunch of questions, but this looks great over all, so most of them are more for my information than something that actually needs changing

AshishDhama commented on actions/views/lhs.js:
```diff
@@ -0,0 +1,45 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import Constants, {ActionTypes, SidebarSize} from 'utils/constants';
```
Is there some specific reason behind this file being `js` instead of `ts`

kyeongsoosoo replied:
Files were mixed in the process of updating the branch. I deleted it now.

AshishDhama commented on components/advanced_text_editor/formatting_bar/formatting_bar.tsx:
```diff
@@ -224,10 +224,12 @@ const FormattingBar = (props: FormattingBarProps): JSX.Element => {
         }
     }, [getCurrentSelection, getCurrentMessage, applyMarkdown, showHiddenControls, toggleHiddenControls, disableControls]);
 
+    const leftPosition = wideMode === 'min' ? (x ?? 0) + 55 : x ?? 0;
```
Can this hardcoded value be avoided `55`

AshishDhama commented on components/advanced_text_editor/formatting_bar/hooks.tsx:
```diff
@@ -34,7 +34,11 @@ const useResponsiveFormattingBar = (ref: React.RefObject<HTMLDivElement>): WideM
+        if (ref.current.clientWidth < 310) {
+            setWideMode('min');
+        }
+    }, 10), []);
```
We should add `ref.current.clientWidth` as a dependency.

AshishDhama commented:
adding @calebroseland As recently worked layouts

kyeongsoosoo committed:
add resizable component

kyeongsoosoo committed:
apply resizable component

kyeongsoosoo committed:
add WindowSizeObserver component

kyeongsoosoo committed:
avoid hard-coded value

kyeongsoosoo committed:
fix test

kyeongsoosoo committed:
Merge branch 'master' of https://github.com/mattermost/mattermost-webapp into MM-27936_resize_sidebar

kyeongsoosoo commented on components/resizable_sidebar/resizable_lhs/index.tsx:
```diff
@@ -0,0 +1,171 @@
+
+    const handleDoubleClick = useCallback(() => {
+        if (lhsRef.current) {
+            lhsRef.current.style.width = `${DEFAULT_LHS_WIDTH}px`;
```
```ts
    const handleMouseMove = useCallback(requestAnimationFrameForMouseMove((e: MouseEvent) => {
       ...
```

The requestAnimationFrame should be used, but there was a problem that the function was recreated when the state outside the function was used.

hmhealey replied:
Oh okay, so if I understand correctly, the problem was: storing the width in state caused the `mousemove` handler to be recreated so it was being fired incorrectly.

Is that right?

kyeongsoosoo replied:
Yes, that's right. That's why I used ref.

kyeongsoosoo commented on components/resizable_sidebar/resizable_lhs/index.tsx:
```diff
@@ -0,0 +1,171 @@
+            return;
+        }
+
+        if (!isResizeLineSelected) {
```
```ts
   useEffect(() => {
        if (!isResizeLineSelected || !dir) {
            return () => {};
        }

        window.addEventListener('mousemove', handleMouseMove);
```

Before registering the event, we are checking to see if the resize has started. Would this be enough?

hmhealey replied:
Yep, I think that should work. That way, we're only registering the `mousemove` handler when the user is dragging the resize line.

We might not actually need to check `dir` there, but I don't think it hurts to include that since that should never actually change

kyeongsoosoo committed:
fix lint

kyeongsoosoo commented on sass/layout/_sidebar-right.scss:
```diff
@@ -5,10 +5,62 @@
+    transition: none !important;
+}
+
+#resizeRHSLine {
```
Now I changed it to use class.

kyeongsoosoo committed:
add ResizeDirection enum

kyeongsoosoo committed:
pass an object to classNames

kyeongsoosoo committed:
create a folder for resizable

kyeongsoosoo commented on components/resizable_sidebar/Resizable.tsx:
```diff
@@ -0,0 +1,253 @@
+    children: React.ReactNode;
+}
+
+export type ResizeDirection = 'left' | 'right'
```
I added an `enum`.

kyeongsoosoo commented on sass/components/_module.scss:
```diff
@@ -52,3 +52,4 @@
 @import 'infinite-scroll';
 @import 'line-switch';
 @import 'apps-modal';
+@import 'resizable';
```
I created a folder for `resizable` and moved `resizable.scss` to this folder.

calebroseland commented on components/window_size_observer/WindowSizeObserver.tsx:
```diff
@@ -0,0 +1,90 @@
+const largeSidebarMediaQuery = window.matchMedia(`(min-width: ${Constants.MEDIUM_SIDEBAR_BREAKPOINT + 1}px) and (max-width: ${Constants.LARGE_SIDEBAR_BREAKPOINT}px)`);
+const xLargeSidebarMediaQuery = window.matchMedia(`(min-width: ${Constants.LARGE_SIDEBAR_BREAKPOINT + 1}px)`);
+
+function WindowSizeObserver() {
```
There should as little as possible JS running when resizing the window.. Could all of this logic be included in or launched from here? https://github.com/mattermost/mattermost-webapp/blob/d0230dc4dde704d69874f4d9e795554c8b824eec/actions/views/browser.ts#L8


This seems extraneous. Could the existing breakpoints based on `views.browser.windowSize` be used for some of this?


<img width="400" alt="image" src="https://user-images.githubusercontent.com/11724372/214870673-68b90b36-b0ca-40e4-a71b-a8f87771eec2.png">

kyeongsoosoo replied:
I thought I should distinguish between `browser.windowSize` and `lhs.size` because breakpoint is applied differently.
To be included in one, I think we need to add more break points to the function `emitBrowserWindowResized`. That would make the function more complicated, would that be okay?

<img width="318" alt="스크린샷 2023-01-31 오전 12 17 58" src="https://user-images.githubusercontent.com/59413880/215517352-d73a5d46-adbc-4af5-8db4-bd273a57d813.png">

calebroseland reviewed (changes requested):
I like the resize handles and the delayed highlight. The resize handles have good boundaries/sizes and are easy to find IMO. 

Where things get quite unstable for me is in the actual resize process: it doesn't stay in-sync with the cursor position as you resize, and gets janky or even stuck in various scenarios. 

The breakpoints don't seem to be set on initial window load, and so you'd have to manually resize before the LHS+RHS breakpoints are properly set. Initial seems hardcoded to `medium`.

Also, I don't think the individual sidebar breakpoints need to be kept in redux. Instead, the resize min/max could be derived by the existing `WindowSizes` breakpoints as stored in `views.browser.windowSize`.

https://user-images.githubusercontent.com/11724372/214876933-baa5c9ae-507e-47d4-a514-14716d116132.mp4


![CleanShot 2023-01-26 at 09 13 04](https://user-images.githubusercontent.com/11724372/214876183-64c9a391-7f80-41cc-9ae9-358a035bceae.gif)


![CleanShot 2023-01-26 at 09 14 33](https://user-images.githubusercontent.com/11724372/214874371-fa2c80f2-8140-4be7-885e-d8b6c530e3ff.gif)

kyeongsoosoo committed:
add rightWidthHolderRef prop

kyeongsoosoo committed:
set size on initial window load

kyeongsoosoo commented:
@calebroseland  When we snap the panel width to the default, the sync between the cursor position and the resizing line becomes out of sync.
In this situation, can you help me which UX would be good to sync again?

https://user-images.githubusercontent.com/59413880/215628024-a5d566ba-3445-43d3-8812-dc6f4852b2de.mov

kyeongsoosoo committed:
do not resize if the cursor is outside the sidebar

kyeongsoosoo committed:
Merge branch 'master' of https://github.com/mattermost/mattermost-webapp into MM-27936_resize_sidebar

kyeongsoosoo commented:
@hmhealey I agree it's not right to name it `ref`. I changed the prop name.

matthewbirtch commented:
> I made breakpoints to be set on the initial window load. And I synchronized the cursor and the sidebar.
> The only problem is that when we snap the panel width to the default, the sync between the cursor position and the resizing line becomes out of sync.
> In this situation, could you help me which UX would be good to synchronize them again?

@kyeongsoosoo can we re-sync the position when the cursor moves > 16px beyond the snap point in either direction? In addition, I wonder if there is anything we can do to detect the speed of the mouse movement and disable the snapping if the user is expending or collapsing the panels quickly. Right now, it just gets stuck on the snap point no matter how quickly you're adjusting the width.

I would recommend playing around with Slack's current implementation for snap points to see what they've done.

kyeongsoosoo committed:
resynchronize the cursor with the line after snap

kyeongsoosoo committed:
snap only when drag speed is slower than limit

AshishDhama commented:
Removing myself as there are already enough dev reviewers

matthewbirtch commented:
> @matthewbirtch I tested the implementation of Slack and updated two things.
> 
> > can we re-sync the position when the cursor moves > 16px beyond the snap point in either direction?
> 
>  2023-02-08.1.19.59.mov 
> > I wonder if there is anything we can do to detect the speed of the mouse movement and disable the snapping if the user is expanding or collapsing the panels quickly.
> 
>  2023-02-08.1.20.20.mov

This is a really great improvement @kyeongsoosoo. Love it! @calebroseland I think this should fix some of the issues you were identifying.

matthewbirtch commented:
> 3\. For the double-click behavior, it looks like it animates back to the default width in the RHS, but not for th LHS. It should be consistent. And I don't think we need the animation at all.

One other thing I just noticed. I thought we fixed the above, but the RHS still seems to animate when I double-click. I think we can just remove the animation and make it snap back to the default width immediately when you double-click.

kyeongsoosoo committed:
remove the animation when double-click

mattermost-build commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

calebroseland commented:
Hi @kyeongsoosoo, I apologize for the delay in getting back to you, and thank you for all the work you've done on this.

I've been testing and experimenting with this, and I think it's close to the finish line. However, there's now a pause on any new features until sometime later this year, so we'll be closing this for now. We will likely reopen the ticket down the road, so there may be a possibility for it later this year.

I and I'm sure others were looking forward to this one, but we'll have to wait and see given the pause. Again, thank you for all your work on this.

matthewbirtch commented:
> Hi @kyeongsoosoo, I apologize for the delay in getting back to you, and thank you for all the work you've done on this.
> 
> I've been testing and experimenting with this, and I think it's close to the finish line. However, there's now a pause on any new features until sometime later this year, so we'll be closing this for now. We will likely reopen the ticket down the road, so there may be a possibility for it later this year.
> 
> I and I'm sure others were looking forward to this one, but we'll have to wait and see given the pause. Again, thank you for all your work on this.

That's really too bad, we were so close here (at least from a UX standpoint)

kyeongsoosoo commented:
@calebroseland @matthewbirtch 
That's too bad😭 
I hope this PR will be useful when ticket is reopened for this feature :)
(If possible, I will participate!)

ghost commented:
Hi @kyeongsoosoo hope you are well! We have migrated to using a monorepo 🎉 and thus the web app code is located here https://github.com/mattermost/mattermost-server

Would you be open to reopening this PR in the new repository? I can assist you with any issues you might have along the way, although setup is quite straightforward!

kyeongsoosoo commented:
Hi, @nevyangelova  
I'll reopen this pr in the new repository!
I think I can do it without difficulty because I have already experienced the migration process.
(https://github.com/mattermost/mattermost-webapp/pull/12256)

ghost commented:
@kyeongsoosoo awesome. Feel free to add me as a reviewer as well :))
